core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,653 @@
+   */
+  public static final Context ROOT = new Context(null);
+
+  /**
+   * Create a {@link Key} with the given name.
+   */
+  public static <T> Key<T> key(String name) {
+    return new Key<T>(name);
+  }
+
+  /**
+   * Create a {@link Key} with the given name and default value.
+   */
+  public static <T> Key<T> keyWithDefault(String name, T defaultValue) {
+    return new Key<T>(name, defaultValue);
+  }
+
+  /**
+   * Return the context associated with the current thread, will never return {@code null} as
+   * the {@link #ROOT} context is implicitly associated with all threads.
+   *
+   * <p>Will never return {@link CancellableContext} even if one is attached, instead a
+   * {@link Context} is returned with the same properties and lifetime. This is to avoid
+   * code stealing the ability to cancel arbitrarily.
+   */
+  public static Context current() {
+    ArrayDeque<Context> stack = contextStack.get();
+    if (stack.isEmpty()) {
+      return ROOT;
+    }
+    return stack.peekLast();
+  }
+
+  private final Context parent;
+  private final Object[][] keyValueEntries;
+  private final boolean cascadesCancellation;
+  private ArrayList<ExecutableListener> listeners;
+  private CancellationListener parentListener = new CancellationListener() {
+    @Override
+    public void cancelled(Context context) {
+      notifyAndClearListeners();
+    }
+  };
+
+  /**
+   * Construct a context that cannot be cancelled and will not cascade cancellation from its parent.
+   */
+  private Context(Context parent) {
+    this.parent = parent;
+    keyValueEntries = EMPTY_ENTRIES;
+    cascadesCancellation = false;
+  }
+
+  /**
+   * Construct a context that cannot be cancelled but will cascade cancellation from its parent if
+   * it is cancellable.
+   */
+  private Context(Context parent, Object[][] keyValueEntries) {
+    this.parent = parent;
+    this.keyValueEntries = keyValueEntries;
+    cascadesCancellation = true;
+  }
+
+  boolean canBeCancelled() {
+    // A context is cancellable if it cascades from its parent and its parent is
+    // cancellable.
+    return (cascadesCancellation && this.parent != null && this.parent.canBeCancelled());
+  }
+
+  /**
+   * Create a new context which is independently cancellable, the new context will cascade
+   * cancellation from its parent.
+   */
+  public CancellableContext withCancellation() {
+    return new CancellableContext(this);
+  }
+
+  /**
+   * Create a new context which will cancel itself after an absolute deadline expressed as
+   * nanoseconds in the {@link System#nanoTime()} clock.
+   *
+   * <p>It is recommended that callers only use this method when propagating a derivative of
+   * a received existing deadline. When establishing a new deadline {@link #withDeadlineAfter}
+   * is the better mechanism.
+   */
+  public CancellableContext withDeadlineNanoTime(long deadlineNanoTime) {
+    return withDeadlineAfter(Math.max(0, deadlineNanoTime - System.nanoTime()),
```

Comment:
The `max` is unnecessary. Negative values are explicitly permitted for `ScheduledExecutorService`:

> Zero and negative delays (but not periods) are also allowed in schedule methods, and are treated as requests for immediate execution.